package/src/main/java/com/azure/cosmos/cassandra/CosmosLoadBalancingPolicy.java

```diff
@@ -0,0 +1,442 @@
+     */
+    @Override
+    @NonNull
+    public Queue<Node> newQueryPlan(@Nullable final Request request, @Nullable final Session session) {
+
+        this.refreshHostsIfDnsExpired();
+
+        final ArrayDeque<Node> queryPlan =  new ArrayDeque<>();
+        final int start = this.index.updateAndGet(value -> value > Integer.MAX_VALUE - 10_000 ? 0 : value + 1);
+
+        if (isReadRequest(request)) {
+            addTo(queryPlan, start, this.readLocalDcNodes);
+        }
+
+        addTo(queryPlan, start, this.writeLocalDcNodes);
+        addTo(queryPlan, start, this.remoteDcNodes);
+
+        return queryPlan;
+    }
+
+    @Override
+    public void onAdd(@NonNull final Node node) {
+        this.onUp(node);
+    }
+
+    @Override
+    public void onDown(@NonNull final Node node) {
+
+        if (node.getDatacenter() == null) {
+            return;
+        }
+
+        if (!this.readDatacenter.isEmpty() && node.getDatacenter().equals(this.readDatacenter)) {
+            this.readLocalDcNodes.remove(node);
+        }
+
+        if (!this.writeDatacenter.isEmpty()) {
+            if (node.getDatacenter().equals(this.writeDatacenter)) {
+                this.writeLocalDcNodes.remove(node);
+            }
+        } else if (Arrays.asList(this.getLocalAddresses()).contains(this.getAddress(node))) {
+            this.writeLocalDcNodes.remove(node);
+        } else {
+            this.remoteDcNodes.remove(node);
+        }
+    }
+
+    @Override
+    public void onRemove(@NonNull final Node node) {
+        this.onDown(node);
+    }
+
+    @Override
+    public void onUp(@NonNull final Node node) {
+
+        if (node.getDatacenter() == null) {
+            return;
+        }
+
+        if (!this.readDatacenter.isEmpty() && node.getDatacenter().equals(this.readDatacenter)) {
+            this.readLocalDcNodes.addIfAbsent(node);
+        }
+
+        if (!this.writeDatacenter.isEmpty()) {
+            if (node.getDatacenter().equals(this.writeDatacenter)) {
+                this.writeLocalDcNodes.addIfAbsent(node);
+            }
+        } else if (Arrays.asList(this.getLocalAddresses()).contains(this.getAddress(node))) {
+            this.writeLocalDcNodes.addIfAbsent(node);
+        } else {
+            this.remoteDcNodes.addIfAbsent(node);
+        }
+    }
+
+    // endregion
+
+    // region Privates
+
+    /**
+     * DNS lookup based on the {@link #globalEndpoint} address.
+     * <p>
+     * If {@link #dnsExpiryTimeInSeconds} has elapsed, the array of local addresses is also updated.
+     *
+     * @return value of {@link #localAddresses} which will have been updated, if {@link #dnsExpiryTimeInSeconds}
+     * has elapsed.
+     *
+     * @throws IllegalStateException if the DNS could not resolve the {@link #globalEndpoint} and local addresses have
+     * not yet been enumerated.
+     */
+    private InetAddress[] getLocalAddresses() {
+
+        if (this.localAddresses == null || this.dnsExpired()) {
+            try {
```

Comment:
Add a Monitor/Semaphore to prevent multiple threads for entering this loop?

Reply:
I changed this to a `synchronized` method on my local branch. I won't resolve this comment now as I'm unclear on the synchronization requirements for load balancing policy objects. I'll look into that before we merge.

Reply:
I've addressed a number of `CosmosLoadBalancingPolicy` concurrency issues related to this issue and our use of `CopyOnWriteArrayList`. I will resolve this issue which we can plan on addressing with PR